# PRACTICE PROBLEMS ON STRING & LIST


**<font color='red'>[MUST MAINTAIN VARIABLE NAMING CONVENTIONS FOR ALL THE TASKS]</font>**
**<font color='red'>[Solve all the tasks sequentially]</font>**


### Task 1
Write a python program that will take a number as an input and check whether the number is Awesome or not. If the number is Awesome, it prints True. Otherwise False.

**Awesome number**: a number where every digit is less than its immediate left digit is called an Awesome number. A single-digit number cannot be an awesome number(e.g. 5421 is an Awesome number.)

==========================================================

**Sample Input1:**\
976321\
**Sample Output1:**\
True

==========================================================

**Sample Input2:**\
9766321\
**Sample Output2:**\
False

==========================================================

**Sample Input3:**\
9\
**Sample Output3:**\
False

==========================================================

**Sample Input4:**\
78\
**Sample Output4:**\
False

==========================================================

**Sample Input5:**\
87\
**Sample Output4:**\
True


In [20]:
#TODO 966321

num_1 = input()
num_2 = num_1[1:]

checker = int(num_1[0])

awesome = True
if len(num_1)>1:
    for i in num_2:
        if int(i)>=checker:
            awesome = False
            break
        checker = int(i)   
    
    if awesome == False:
        print("False")
    if awesome == True:
        print("True")
           
else:
    print("False")


979210
False


### Task 2

Write a python program that will take a number a list of numbers(integer) as an input from the user and print whether the number is Awesome or not. 

==========================================================

**Sample Input1:**\
[976321, 321, 9763221, 9742, 876, 3211]\
**Sample Output1:**\
976321 is an awesome number.\
321 is an awesome number.\
9763221 is a not-so-awesome number.\
9742 is an awesome number.\
876 is an awesome number.\
3211 is a not-so-awesome number.

==========================================================

**Sample Input2:**\
[97821, 97210, 979210]\
**Sample Output2:**\
97821 is a not-so-awesome number.\
97210 is an awesome number.\
979210 is a not-so-awesome number.


In [16]:
#Todo
list_1 = input()[1:-1].split(", ")

for i in list_1:
    checker = int(i[0])
    counter = 0
    flag = True
    for j in i:
        if j == i[0] and counter== 0:
            counter+=1
            continue
        if int(j)>=checker:
            print(i, "is a not-so-awesome number.")
            flag = False
            break
        checker= int(j) 
    if flag!= False:
        print(i, "is an awesome number.")

[97821, 97210, 979210]
97821 is a not-so-awesome number.
97210 is an awesome number.
979210 is a not-so-awesome number.


### Task 3


Write a python program that will take a string as input from the user where each number is separated by a comma. Then gets the numbers from the given string and removes all extra spaces. Stores the clean numbers in a list and prints it.

==========================================================

**Hints:** For obtaining the numbers from the string, use split(). For cleaning the data, use strip().

==========================================================

**Sample Input1:**\
"97821, 1 &nbsp;&nbsp;&nbsp;  , 97210,&nbsp;&nbsp; 963, &nbsp;&nbsp;&nbsp;979210 , 979210 "\
**Sample Output1:**\
Data after cleaning: [97821, 1, 97210, 963, 979210, 979210]

==========================================================

**Sample Input2:**\
"976321, 321&nbsp;&nbsp;,&nbsp;&nbsp;&nbsp;9763221, &nbsp;&nbsp;&nbsp;9742, 876,&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3211"\
**Sample Output2:**\
Data after cleaning: [976321, 321, 9763221, 9742, 876, 3211]

==========================================================

**Sample Input3:**\
"976321, 321 ,&nbsp;&nbsp;&nbsp; 9763221, 9742, &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;876, &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3211, 976321&nbsp;&nbsp;&nbsp;&nbsp;, 9742"\
**Sample Output3:**\
Data after cleaning: [976321, 321, 9763221, 9742, 876, 3211, 976321, 9742]


In [17]:
#TODO
list_1 = input()[1:-1].split(",")
list_2 = []
str_1 = ""

for i in list_1:
    for j in i:
        if j == " ":
            continue
        else:
            str_1+=j
            list_2.append(int(j))
print(list_2)            

"976321, 321  ,   9763221,    9742, 876,        3211"
[9, 7, 6, 3, 2, 1, 3, 2, 1, 9, 7, 6, 3, 2, 2, 1, 9, 7, 4, 2, 8, 7, 6, 3, 2, 1, 1]


### Task 4

Write a python program that will take a list of numbers(integer) as an input from the user and does the following.
* counts the number of valid, invalid, and duplicate numbers and prints these to the user.
* stores the valid and invalid numbers in a list format and prints these to the user. You need to make sure that a number is stored only Once. **Valid numbers:** A number with a minimum length of 4 and a maximum length of 6 is considered to be valid for this assignment.
* Calculate and print the percentage of valid numbers, invalid numbers, and duplicate numbers to the user. 

==========================================================

**Hints:** Use membership operator in and not in for checking the duplicates. For calculating the percentage, count the numbers of a particular number type.

==================================================


**Sample Input1:**\
[97821, 1, 97210, 963, 979210, 979210]\
**Sample Output1:**\
================================================\
Total numbers present in the given string: 6\
Valid number count: 3\
Valid numbers in list: [97821, 97210, 979210]\
Invalid number count: 2\
Invalid numbers in list: [1, 963]\
Duplicate number count: 1\
================================================\
Valid number percentage(%): 50.0\
Invalid number percentage(%): 33.333333333333336\
Duplicate number percentage(%): 16.666666666666668\
================================================\


**Sample Input2:**\
[976321, 321, 9763221, 9742, 876, 3211]\
**Sample Output2:**\
================================================\
Total numbers present in the given string: 6\
Valid number count: 3\
Valid numbers in list: [976321, 9742, 3211]\
Invalid number count: 3\
Invalid numbers in list: [321, 9763221, 876]\
Duplicate number count: 0\
================================================\
Valid number percentage(%): 50.0\
Invalid number percentage(%): 50.0\
Duplicate number percentage(%): 0.0\
================================================\

**Sample Input3:**\
[976321, 321, 9763221, 9742, 876, 3211, 976321, 9742]\
**Sample Output3:**\
================================================\
Total numbers present in the given string: 8\
Valid number count: 3\
Valid numbers in list: [976321, 9742, 3211]\
Invalid number count: 3\
Invalid numbers in list: [321, 9763221, 876]\
Duplicate number count: 2\
================================================\
Valid number percentage(%): 37.5\
Invalid number percentage(%): 37.5\
Duplicate number percentage(%): 25.0\
================================================\


976321, 321, 9763221, 9742, 876, 3211, 976321, 9742
Total numbers present in the given string:  8
Valid number count:  3
Valid numbers in list:  ['976321', '9742', '3211']
Invalid number count:  3
Invalid numbers in list:  ['321', '9763221', '876']
Duplicate number count:  2
Valid number percentage:  37.5 %
Invalid number percentage:  37.5 %
Duplicate number percentage:  25.0 %


### Task 5

Write a python program that will take a list of numbers(integer) as an input from the user. Then stores the valid numbers, the sum of digits, and multiplication of the digits for each valid number in a list where each number is the 1st element of the list, the sum of digits is the 2nd element of the list and the multiplication of digits is the 3rd element of the list. Finally, print the nested list to the user. 

==========================================================

**Sample Input1:**\
[97821, 97210, 979210]\
**Sample Output1:**\
Data [number, sum of digits, product of digits]: [[97821, 27, 1008], [97210, 19, 0], [979210, 28, 0]]

================================================

**Sample Input2:**\
[976321, 9742, 3211, 30987]\
**Sample Output2:**\
Data [number, sum of digits, product of digits]: [[976321, 28, 2268], [9742, 22, 504], [3211, 7, 6], [30987, 27, 0]]

================================================

**Sample Input3:**\
[976321, 9640, 3211]\
**Sample Output3:**\
Data [number, sum of digits, product of digits]: [[976321, 28, 2268], [9640, 19, 0], [3211, 7, 6]]

================================================


### Task 6 (Merger of Task1-Task5)

Write a python program that will take a string as input from the user where each number is separated by a comma. The numbers must have a minimum length of 4 and a maximum length of 6 to be considered valid for this assignment. Then, you need to complete the following. 

* Part1: get the numbers from the given string, remove all extra spaces. (clean the data). Print the clean data to the user.
* Part2:count the number of valid, invalid, and duplicate numbers and print these to the user.
* Part3: store the valid and invalid numbers in a list format and print these to the user. You need to make sure that a number is stored only Once.
* Part4: Calculate and print the percentage of valid numbers, invalid numbers, and duplicate numbers to the user.
* Part5:store the valid numbers, the sum of digits, and multiplication of the digits for each valid number in a list where each number is the 1st element of the list, the sum of digits is the 2nd element of the list and the multiplication of digits is the 3rd element of the list. Finally, print the nested list to the user. 
* Part6: Check whether the valid numbers are Awesome or not. **Awesome number:** a number where every digit is less than its immediate left digit is called is an awesome number. (e.g. 5421 is an awesome number).

==============================================================================================

**Sample Input1:**\
"97821, 1 &nbsp;&nbsp;&nbsp;  , 97210,&nbsp;&nbsp; 963, &nbsp;&nbsp;&nbsp;979210 , 979210 "\
**Sample Output1:**\
Data after cleaning: [97821, 1, 97210, 963, 979210, 979210]\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
================================================\
Total numbers present in the given string: 6\
Valid number count: 3\
Valid numbers in list: [97821, 97210, 979210]\
Invalid number count: 2\
Invalid numbers in list: [1, 963]\
Duplicate number count: 1\
================================================\
Valid number percentage(%): 50.0\
Invalid number percentage(%): 33.333333333333336\
Duplicate number percentage(%): 16.666666666666668\
================================================\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
Data of valid numbers:\
[number, sum of digits, product of digits]:\
[[976321, 28, 2268], [9640, 19, 0], [3211, 7, 6]]\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
^_^ Awesomeness checking of the Valid numbers ^_^\
976321 is an awesome number.\
9640 is an awesome number.\
3211 is a not-so-awesome number.\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

==============================================================================================


**Sample Input2:**\
"976321, 321&nbsp;&nbsp;,&nbsp;&nbsp;&nbsp;9763221, &nbsp;&nbsp;&nbsp;9742, 876,&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3211"\
**Sample Output2:**\
Data after cleaning: [976321, 321, 9763221, 9742, 876, 3211]\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
================================================\
Total numbers present in the given string: 6\
Valid number count: 3\
Valid numbers in list: [976321, 9742, 3211]\
Invalid number count: 3\
Invalid numbers in list: [321, 9763221, 876]\
Duplicate number count: 0\
================================================\
Valid number percentage(%): 50.0\
Invalid number percentage(%): 50.0\
Duplicate number percentage(%): 0.0\
================================================\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
Data of valid numbers:\
[number, sum of digits, product of digits]:\
[[976321, 28, 2268], [9640, 19, 0], [3211, 7, 6]]\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
^_^ Awesomeness checking of the Valid numbers ^_^\
976321 is an awesome number.\
9640 is an awesome number.\
3211 is a not-so-awesome number.\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX


==============================================================================================

**Sample Input3:**\
"976321, 321 ,&nbsp;&nbsp;&nbsp; 9763221, 9742, &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;876, &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3211, 976321&nbsp;&nbsp;&nbsp;&nbsp;, 9742"\
**Sample Output3:**\
Data after cleaning: [976321, 321, 9763221, 9742, 876, 3211, 976321, 9742]\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
================================================\
Total numbers present in the given string: 8\
Valid number count: 3\
Valid numbers in list: [976321, 9742, 3211]\
Invalid number count: 3\
Invalid numbers in list: [321, 9763221, 876]\
Duplicate number count: 2\
================================================\
Valid number percentage(%): 37.5\
Invalid number percentage(%): 37.5\
Duplicate number percentage(%): 25.0\
================================================\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
Data of valid numbers:\
[number, sum of digits, product of digits]:\
[[976321, 28, 2268], [9640, 19, 0], [3211, 7, 6]]\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\
^_^ Awesomeness checking of the Valid numbers ^_^\
976321 is an awesome number.\
9640 is an awesome number.\
3211 is a not-so-awesome number.\
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

In [3]:
s = "1+2+3+4"
h =str(1+2+3+4)
print((h))
print(type(h))

10
<class 'str'>


In [ ]:
#TODO

In [3]:
a = [1,2,2,2,2,2,3,4,5]

a1 = a.copy() #[ a1 = a Don't use]
print(a1)

a[2] = "heda"
print(a)

print(a.index(2))

print(a.count(2))

# print(a.reverse(3,4))

a.append(999)
print(a)

a.insert(2,999)
print(a)

a.pop()
print(a)

a.pop(1) #[index]
print(a)

# a.remove("heda") #[character]
# print(a)

# del a[7]
# print(a)

# a.clear()
# print(a)

# list_0 = ["stoic bliss"]
# list_1 = [23,98,77,53,44]
# list_2 = [11, 10, 9, 8, 7]
# list_3 = list_1 + list_2
# print(list_3)

# list_3.sort()
# print(list_3)

# list_3.reverse()
# print(list_3)

# list_3.extend(list_0)
# print(list_3)

# list_3.extend(["shelby"])
# print(list_3)

# list_3.reverse()
# print(list_3)

[1, 2, 2, 2, 2, 2, 3, 4, 5]
[1, 2, 'heda', 2, 2, 2, 3, 4, 5]
1
4
[1, 2, 'heda', 2, 2, 2, 3, 4, 5, 999]
[1, 2, 999, 'heda', 2, 2, 2, 3, 4, 5, 999]
[1, 2, 999, 'heda', 2, 2, 2, 3, 4, 5]
[1, 999, 'heda', 2, 2, 2, 3, 4, 5]


In [107]:
j = []

for i in range(1,11):
    j.append(i)
    print(j)

[1]
[1, 2]
[1, 2, 3]
[1, 2, 3, 4]
[1, 2, 3, 4, 5]
[1, 2, 3, 4, 5, 6]
[1, 2, 3, 4, 5, 6, 7]
[1, 2, 3, 4, 5, 6, 7, 8]
[1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [8]:
capt_dict = {"Portugal":"Ronaldo", "Argentina":"Messi", "Spain": "Ramos"}
print(capt_dict["Portugal"])

print(capt_dict.get("Portugal"))
print(capt_dict["Brazil"])

Ronaldo
Ronaldo


KeyError: 'Brazil'

In [15]:
capt_dict = {"Portugal":"Ronaldo", "Argentina":{"Name":"Messi", "Goal":"22"}, "Spain": "Ramos"}
print(capt_dict.get("Argentina"))

print(capt_dict["Argentina"]["Name"])
# print(capt_dict.get("Argentina")("Name")) = error
capt_dict.update({"Spain":"Neymar"})
print(capt_dict)
capt_dict.update({"Brasil":"Neymar"})
print(capt_dict)

capt_dict["Spain"] = "Michiko"
print(capt_dict)

# capt_dict.clear()
# print(capt_dict) ---> {}

remove = capt_dict.popitem()
print(remove)
print(capt_dict)

remove = capt_dict.pop("Spain")
print(remove)
print(capt_dict)

del capt_dict["Portugal"]
print(capt_dict)

# del capt_dict
# print(capt_dict)



{'Name': 'Messi', 'Goal': '22'}
Messi
{'Portugal': 'Ronaldo', 'Argentina': {'Name': 'Messi', 'Goal': '22'}, 'Spain': 'Neymar'}
{'Portugal': 'Ronaldo', 'Argentina': {'Name': 'Messi', 'Goal': '22'}, 'Spain': 'Neymar', 'Brasil': 'Neymar'}
{'Portugal': 'Ronaldo', 'Argentina': {'Name': 'Messi', 'Goal': '22'}, 'Spain': 'Michiko', 'Brasil': 'Neymar'}
('Brasil', 'Neymar')
{'Portugal': 'Ronaldo', 'Argentina': {'Name': 'Messi', 'Goal': '22'}, 'Spain': 'Michiko'}
Michiko
{'Portugal': 'Ronaldo', 'Argentina': {'Name': 'Messi', 'Goal': '22'}}
{'Argentina': {'Name': 'Messi', 'Goal': '22'}}


In [67]:
capt_dict = {"Portugal":"Ronaldo", "Argentina":{"Name":"Messi", "Goal":"22"}, "Spain": "Ramos"}
hi = (capt_dict.keys())
print(hi)

bye = capt_dict.values()
print(bye)

hola = capt_dict.items()
print(hola)
print(type(hi), type(bye), type(hola))

# dir(dict)

for i in capt_dict:
    print(i, end = " ")
print()

for i in capt_dict.keys():
    print(i, end = " ")    
print()

for i in capt_dict.values():
    print(i, end = " ") 
print()

for i in capt_dict.items():
    print(i, end = " ")      
print()

capt_dict.update({"Argentina":"Messi"})
for i,j in capt_dict.items():
    print(j, "is the captain of", i)      
    
print(len(capt_dict))   

dada = max(capt_dict)
print(dada)

print(min(capt_dict.values()))

print(sorted(capt_dict))

print(capt_dict)

dict_keys(['Portugal', 'Argentina', 'Spain'])
dict_values(['Ronaldo', {'Name': 'Messi', 'Goal': '22'}, 'Ramos'])
dict_items([('Portugal', 'Ronaldo'), ('Argentina', {'Name': 'Messi', 'Goal': '22'}), ('Spain', 'Ramos')])
<class 'dict_keys'> <class 'dict_values'> <class 'dict_items'>
Portugal Argentina Spain 
Portugal Argentina Spain 
Ronaldo {'Name': 'Messi', 'Goal': '22'} Ramos 
('Portugal', 'Ronaldo') ('Argentina', {'Name': 'Messi', 'Goal': '22'}) ('Spain', 'Ramos') 
Ronaldo is the captain of Portugal
Messi is the captain of Argentina
Ramos is the captain of Spain
3
Spain
Messi
['Argentina', 'Portugal', 'Spain']
{'Portugal': 'Ronaldo', 'Argentina': 'Messi', 'Spain': 'Ramos'}


In [73]:
a = {'Portugal': 'Ronaldo', 'Argentina': 'Messi', 'spain': 'Ramos'}
print(max(a.keys()))

spain


In [2]:
my_tuple = 4+9+21, 7.9, "yo_bruh"

print(my_tuple)
print(type(my_tuple))

a, b, shelby = my_tuple
print(a)
print(shelby)

x = (my_tuple[-1][-1])
print(x)

print(type(x))

print(my_tuple+("yoyo",98.6))

my_tuple_1 = (99,33,3,"yo",["hexa",177,991.01],{"brac",22101,0.820})
print(my_tuple_1)

my_tuple_1[4].append(.7070)
print(my_tuple_1)

# my_dict = {99,33,3,"brac",22101,0.820}
# my_dict_1 = {"bike":"gixxer"}
# print(my_dict)
# print(type(my_dict))
# print(type(my_dict_1))

(34, 7.9, 'yo_bruh')
<class 'tuple'>
34
yo_bruh
h
<class 'str'>
(34, 7.9, 'yo_bruh', 'yoyo', 98.6)
(99, 33, 3, 'yo', ['hexa', 177, 991.01], {0.82, 'brac', 22101})
(99, 33, 3, 'yo', ['hexa', 177, 991.01, 0.707], {0.82, 'brac', 22101})


In [141]:
empty_tuple = ()

player_1 = ["Messi",("Argentina", "FCB")]
print(type(player_1[1]))
# print(type(player_1))

player_1 = ("Messi",("Argentina", "FCB"))
print(player_1[1])
print(player_1)

for i in player_1:
    print(i, end = " ")
print()

frnds = "ST", "Mentor", "Advisor" 
print(frnds)

<class 'tuple'>
('Argentina', 'FCB')
('Messi', ('Argentina', 'FCB'))
Messi ('Argentina', 'FCB') 
('ST', 'Mentor', 'Advisor')


In [3]:
def hahaa():
    print("whassup polapain")
hahaa() 

def yoyo(player,sports):
    print(player, "plays", sports)
yoyo("Adnan","Football")    

def name(*player):
    print("xnxx", player[0])
name("Messi", "Ronaldo", "Neymar")    

def kids (child1, child2, child3):
    print("The name of first kid is", child3)
kids(child2 = "Mokhles", child1 = "Mofiz", child3 = "Rabon")    

whassup polapain
Adnan plays Football
xnxx Messi
The name of first kid is Rabon


In [7]:

# initializing dictionary
test_dict = {'Gfg': 4, 'is': 1, 'best': 8, 'for': 10, 'geeks': 9}
  
# # printing original dictionary
# print("The original dictionary is : " + str(test_dict))
  
# # initializing swap indices
# i, j = 1, 3
  
# # conversion to tuples
tups = list(test_dict.items())
print(type(tups[0][1]))
  
# # swapping by indices
# tups[i], tups[j] = tups[j], tups[i]
  
# # converting back
# res = dict(tups)
  
# # printing result
# print("The swapped dictionary : " + str(res))

<class 'int'>
